File: python/stack_hello.py
#!/usr/bin/env python3
"""
Stack-based Hello World - Turing's Bury/Unbury in Python
High-level demonstration showing Python's approach to stacks
"""

def stack_hello_simple():
    """Simple demonstration using Python list as stack"""
    stack = []  # Python list = dynamic array with stack operations
    message = "Hello World!"
    
    print("BURYING (pushing to stack):")
    for char in message:
        stack.append(char)  # append = push
        print(f"  Push '{char}' → Stack: {stack}")
    
    print("\nUNBURYING (popping from stack):")
    result = []
    while stack:
        char = stack.pop()  # pop = built-in!
        print(f"  Pop  '{char}' ← Stack: {stack if stack else '(empty)'}")
        result.append(char)
    
    print(f"\nReversed order: {''.join(result)}")
    print(f"Original order: {''.join(reversed(result))}")


def stack_hello_explicit():
    """More explicit demonstration with custom Stack class"""
    
    class Stack:
        def __init__(self):
            self.items = []
        
        def push(self, item):
            self.items.append(item)
        
        def pop(self):
            if not self.is_empty():
                return self.items.pop()
            return None
        
        def is_empty(self):
            return len(self.items) == 0
        
        def size(self):
            return len(self.items)
        
        def peek(self):
            if not self.is_empty():
                return self.items[-1]
            return None
        
        def __str__(self):
            return str(self.items)
    
    print("\n" + "="*60)
    print("EXPLICIT STACK CLASS DEMONSTRATION")
    print("="*60)
    
    stack = Stack()
    message = "Hello World!"
    
    print(f"\nPushing '{message}' character by character:")
    for i, char in enumerate(message, 1):
        stack.push(char)
        print(f"  {i:2}. Push '{char}' → Stack size: {stack.size()}, Top: '{stack.peek()}'")
    
    print(f"\n Current stack: {stack}")
    
    print("\nPopping all characters:")
    buffer = []
    count = 1
    while not stack.is_empty():
        char = stack.pop()
        buffer.append(char)
        top = stack.peek()
        print(f"  {count:2}. Pop  '{char}' → Stack size: {stack.size()}, " +
              f"Top: '{top if top else 'None'}'")
        count += 1
    
    print(f"\nFinal result (reversed): {''.join(buffer)}")
    print(f"Corrected: {''.join(reversed(buffer))}")


def demonstrate_python_features():
    """Show Python-specific stack features"""
    print("\n" + "="*60)
    print("PYTHON-SPECIFIC FEATURES")
    print("="*60)
    
    # Using collections.deque for efficient stack
    from collections import deque
    
    print("\n1. Using collections.deque (optimized for stack/queue):")
    stack = deque()
    for char in "ABC":
        stack.append(char)
        print(f"   Pushed '{char}' → {list(stack)}")
    
    while stack:
        print(f"   Popped '{stack.pop()}' ← {list(stack) if stack else '(empty)'}")
    
    # List slicing
    print("\n2. Python list slicing (no explicit loop needed):")
    message = "Hello World!"
    print(f"   Original: {message}")
    print(f"   Reversed: {message[::-1]}")
    
    # List comprehension
    print("\n3. List comprehension for stack operations:")
    chars = list("Hello")
    print(f"   Original: {chars}")
    reversed_chars = [chars.pop() for _ in range(len(chars.copy()))]
    print(f"   After pops: {reversed_chars}")
    
    # Context manager could be used for stack-based resource management
    print("\n4. Stack-based resource management (context managers):")
    print("   with open('file') as f:")
    print("       # File handle pushed onto resource stack")
    print("       # Automatically popped/closed on exit")


if __name__ == "__main__":
    print("╔═══════════════════════════════════════════════════════════╗")
    print("║  Python Stack Operations - High-Level Abstraction Demo   ║")
    print("╚═══════════════════════════════════════════════════════════╝")
    
    stack_hello_simple()
    stack_hello_explicit()
    demonstrate_python_features()
    
    print("\n" + "="*60)
    print("KEY PYTHON INSIGHTS:")
    print("="*60)
    print("• Lists have built-in push (append) and pop operations")
    print("• No manual memory management needed")
    print("• Dynamic typing - stack can hold any type")
    print("• Multiple ways to reverse: [::-1], reversed(), manual pop")
    print("• collections.deque more efficient for frequent push/pop")
    print("• Stack concept used in language itself (call stack, 'with')")
    print("="*60)

File: rust/stack_hello.rs
// Stack-based Hello World in Rust
// Demonstrates Rust's ownership system and memory safety

use std::fmt;

// Custom Stack implementation with generic types
struct Stack<T> {
    items: Vec<T>,
}

impl<T> Stack<T> {
    fn new() -> Self {
        Stack { items: Vec::new() }
    }
    
    fn push(&mut self, item: T) {
        self.items.push(item);
    }
    
    fn pop(&mut self) -> Option<T> {
        self.items.pop()
    }
    
    fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
    
    fn size(&self) -> usize {
        self.items.len()
    }
    
    fn peek(&self) -> Option<&T> {
        self.items.last()
    }
}

impl<T: fmt::Display> Stack<T> {
    fn display(&self) {
        print!("[");
        for (i, item) in self.items.iter().enumerate() {
            if i > 0 { print!(", "); }
            print!("{}", item);
        }
        print!("]");
    }
}

fn stack_hello_basic() {
    println!("=== Basic Stack Operations ===\n");
    
    let mut stack = Stack::new();
    let message = "Hello World!";
    
    println!("BURYING (pushing to stack):");
    for ch in message.chars() {
        stack.push(ch);
        print!("  Push '{}' → Stack: ", ch);
        stack.display();
        println!();
    }
    
    println!("\nUNBURYING (popping from stack):");
    let mut buffer = Vec::new();
    while !stack.is_empty() {
        if let Some(ch) = stack.pop() {
            print!("  Pop  '{}' ← Stack: ", ch);
            if stack.is_empty() {
                print!("(empty)");
            } else {
                stack.display();
            }
            println!();
            buffer.push(ch);
        }
    }
    
    let reversed: String = buffer.iter().collect();
    let original: String = buffer.iter().rev().collect();
    
    println!("\nReversed: {}", reversed);
    println!("Original: {}", original);
}

fn demonstrate_ownership() {
    println!("\n{}", "=".repeat(60));
    println!("RUST OWNERSHIP DEMONSTRATION");
    println!("{}", "=".repeat(60));
    
    println!("\n1. Stack owns its data:");
    let mut stack1 = Stack::new();
    stack1.push(String::from("Hello"));
    stack1.push(String::from("World"));
    
    println!("  stack1 contains Strings (heap-allocated)");
    println!("  When stack1 goes out of scope, Strings are freed");
    
    println!("\n2. Moving ownership:");
    let mut stack2 = stack1;  // Ownership transferred!
    // stack1 is now invalid - can't use it!
    println!("  stack1 moved to stack2");
    println!("  Trying to use stack1 now would cause compile error");
    
    if let Some(word) = stack2.pop() {
        println!("  Popped from stack2: {}", word);
    }
    
    println!("\n3. Borrowing (immutable):");
    let stack3: Stack<i32> = Stack::new();
    let size = stack3.size();  // Immutable borrow
    println!("  Borrowed stack3 to check size: {}", size);
    println!("  stack3 still valid after borrow");
    
    println!("\n4. Borrowing (mutable):");
    let mut stack4 = Stack::new();
    stack4.push(42);  // Mutable borrow
    println!("  Mutably borrowed stack4 to push");
    println!("  Can only have ONE mutable borrow at a time!");
}

fn demonstrate_generic_stack() {
    println!("\n{}", "=".repeat(60));
    println!("GENERIC STACK - MULTIPLE TYPES");
    println!("{}", "=".repeat(60));
    
    // Stack of integers
    println!("\n1. Stack<i32>:");
    let mut int_stack = Stack::new();
    for i in 1..=5 {
        int_stack.push(i);
    }
    print!("  ");
    int_stack.display();
    println!();
    
    // Stack of strings
    println!("\n2. Stack<&str>:");
    let mut str_stack = Stack::new();
    str_stack.push("first");
    str_stack.push("second");
    str_stack.push("third");
    print!("  ");
    str_stack.display();
    println!();
    
    // Stack of tuples
    println!("\n3. Stack<(i32, &str)>:");
    let mut tuple_stack = Stack::new();
    tuple_stack.push((1, "one"));
    tuple_stack.push((2, "two"));
    println!("  Stack of tuples: size = {}", tuple_stack.size());
    
    println!("\nRust's generics are zero-cost abstractions!");
    println!("No runtime overhead like dynamic typing");
}

fn demonstrate_safety() {
    println!("\n{}", "=".repeat(60));
    println!("RUST SAFETY GUARANTEES");
    println!("{}", "=".repeat(60));
    
    println!("\n✓ No null pointers - uses Option<T>");
    let mut stack: Stack<i32> = Stack::new();
    match stack.pop() {
        Some(value) => println!("  Got: {}", value),
        None => println!("  Stack empty (None) - safe!"),
    }
    
    println!("\n✓ No buffer overflows - bounds checked");
    println!("  Vec<T> automatically resizes");
    println!("  Index access checked at runtime");
    
    println!("\n✓ No use-after-free");
    println!("  Ownership system prevents dangling pointers");
    println!("  Compiler enforces at compile time!");
    
    println!("\n✓ No data races");
    println!("  One mutable OR multiple immutable borrows");
    println!("  Enforced at compile time!");
    
    println!("\n✓ No manual memory management");
    println!("  RAII - Resource Acquisition Is Initialization");
    println!("  Automatic cleanup when out of scope");
}

fn main() {
    println!("╔═══════════════════════════════════════════════════════════╗");
    println!("║   Rust Stack Operations - Memory Safety & Ownership      ║");
    println!("╚═══════════════════════════════════════════════════════════╝\n");
    
    stack_hello_basic();
    demonstrate_ownership();
    demonstrate_generic_stack();
    demonstrate_safety();
    
    println!("\n{}", "=".repeat(60));
    println!("KEY RUST INSIGHTS:");
    println!("{}", "=".repeat(60));
    println!("• Ownership system prevents memory bugs at compile time");
    println!("• Zero-cost abstractions - as fast as C");
    println!("• Option<T> instead of null pointers");
    println!("• Generics with no runtime overhead");
    println!("• Move semantics by default (no implicit copies)");
    println!("• Borrowing rules enforced by compiler");
    println!("• Memory safety WITHOUT garbage collection");
    println!("{}", "=".repeat(60));
}
